### PR TITLE
change(web): remove support for es5 🏗️

### PR DESCRIPTION
**Pull request by ermshiperete**
Fixes: #11878

@keymanapp-test-bot skip

### CODE REVIEW

File: common/web/keyboard-processor/build.sh [hunk 1]
```diff
@@ -45,21 +45,26 @@ function do_build() {
   tsc --build "$THIS_SCRIPT_PATH/tsconfig.all.json"
 
   # Base product - the main keyboard processor
-  $BUNDLE_CMD    "${KEYMAN_ROOT}/common/web/keyboard-processor/build/obj/index.js" \
+  ${BUNDLE_CMD}    "${KEYMAN_ROOT}/common/web/keyboard-processor/build/obj/src/index.js" \
```
Comment (jahorton) [inline common/web/keyboard-processor/build.sh:48]: Why the change in path here?

Comment (jahorton) [reply common/web/keyboard-processor/build.sh:48]: Also, just to be clear - be _very_ careful when changing the build-output paths like this - it has a tendency to have unwanted side effects in the end-result sourcemap files, often causing the pathing structure shown when debugging via Chrome to become very messy and thus more difficult to utilize.

File: common/web/keyboard-processor/tests/dom/cases/domKeyboardLoader.spec.ts [hunk 2]
```diff
@@ -15,8 +15,8 @@ const device: DeviceSpec = {
 
 describe('Keyboard loading in DOM', function() {
   afterEach(() => {
-    if(window['KeymanWeb']) {
-      window['KeymanWeb'].uninstall();
+    if(window['KeymanWeb' as any]) {
+      (window['KeymanWeb' as any] as any).uninstall();
```
Comment (jahorton) [inline common/web/keyboard-processor/tests/dom/cases/domKeyboardLoader.spec.ts:19]: Try using this at the file's top:

```typescript
declare let KeymanWeb: KeyboardInterface
```

It's possible to declare globals with type, after which you won't need the `as any` stuff.

We already use this pattern in a few places, such as the spot below:

https://github.com/keymanapp/keyman/blob/b15dcef5262df8e92851cdc62a7c00b910755b16/web/src/app/ui/kmwuitoggle.ts#L9

With that in place, it doesn't fix the `window['keyman']` pattern, but it does allow us to just assume that `keyman` exists as a global, which achieves what we're really after anyway.

File: common/web/keyboard-processor/tests/dom/web-test-runner.config.mjs [hunk 3]
```diff
@@ -21,7 +21,7 @@ export default {
   concurrency: 10,
   nodeResolve: true,
   files: [
-    '**/*.spec.ts'
+    'build/lib/**/*.spec.mjs'
```
Comment (jahorton) [inline common/web/keyboard-processor/tests/dom/web-test-runner.config.mjs:24]: I would strongly prefer a specific build folder for tests, then a specific folder for each test set.

Something like `build/test/dom/**/*.spec.mjs` - this pattern is being maintained for `/web/src/test/auto/dom/` tests, so why not be consistent with the pattern and reuse it here?

(While this module's headless tests aren't yet in TS, they could then go under `build/test/headless/**`.)

A general pattern is for `lib` folders to hold bundled / fully-compiled outputs, not testing files... right?  Or am I just mistaken about that?  Either way, that's a pattern I've held to strongly with the lower-level packages; test file outputs do not belong in that folder, in my opinion.  Too messy.

File: web/src/test/auto/dom/web-test-runner.config.mjs [hunk 4]
```diff
@@ -31,34 +31,34 @@ export default {
       name: 'engine/attachment',
       // Relative, from the containing package.json
       files: [
-        'src/test/auto/dom/cases/attachment/**/*.spec.html',
-        'src/test/auto/dom/cases/attachment/**/*.spec.ts'
+        'build/test/dom/cases/attachment/**/*.spec.html',
+        'build/test/dom/cases/attachment/**/*.spec.mjs'
```
Comment (jahorton) [inline web/src/test/auto/dom/web-test-runner.config.mjs:35]: To compare with the comment I made for keyboard-processor tests:  note how the `build/test/dom/**` pattern is being used here, with minimal changes to boot.

File: common/web/lm-worker/build-polyfiller.js
Comment (jahorton) [inline common/web/lm-worker/build-polyfiller.js:1]: Note that this file exists to polyfill for things that would be otherwise missing with ES5, etc.  There's a strong chance we should be able to eliminate this script entirely, though it'd take a bit of extra work.

File: common/web/tslib/README.md
Comment (jahorton) [inline common/web/tslib/README.md:1]: Ooh, it would be nice to drop this if viable.  I'll want to double-check the outputs to ensure that we aren't getting duplicated/triplicated tslib helper functions due to it, just to be sure, but the comments originally at the top here suggest that we should be fine.

The Web filesize-increase reporter should signal a significant spike in filesize if it's not resolving the helpers properly.

File: common/web/types/src/deps/xml2js/parser.js [hunk 5]
```diff
@@ -323,7 +323,11 @@ export class Parser extends EventEmitter {
       } catch (error1) {
         err = error1;
         if (!(this.saxParser.errThrown || this.saxParser.ended)) {
-          this.emit('error', err);
+          if(this.listenerCount('error') > 0) {
+            this.emit('error', err);
+          } else {
+            throw err;
+          }
```
Comment (jahorton) [inline common/web/types/src/deps/xml2js/parser.js:330]: For other reviewers, just in case:

We use `eventemitter3` so that Web can safely compile against common/web/types; `events` is a Node-specific package that won't exist in the browser.  There is _one_ particular point of divergence between the two types:

- In `events`, `.emit('error')` will actually _`throw`_ the error if there are no registered listeners.
- In `eventemitter3`, this specialized behavior is omitted.  It will never `throw` on `.emit('error')` regardless of listener count.

This is the one spot in `xml2js` that uses `.emit('error')`, so we opted for a precise adjustment to accommodate for the difference between the two implementations that will replicate the behavior of the original.  One of the `common/web/types` unit tests fails otherwise due to reliance on the pattern.

File: web/build.sh [hunk 6]
```diff
@@ -39,11 +39,34 @@ builder_describe "Builds engine modules for Keyman Engine for Web (KMW)." \
...
+config=release
+if builder_is_debug_build; then
+  config=debug
+fi
+
+builder_describe_outputs \
+  configure                     "/node_modules" \
+  build                         "/web/build/test/dom/cases/attachment/outputTargetForElement.spec.html" \
+  build:app/browser             "/web/build/app/browser/lib/index.mjs" \
+  build:app/WebViews            "/web/build/app/webview/${config}/keymanweb-webview.js" \
```
Comment (jahorton) [inline web/build.sh:53]:
```suggestion
  build:app/webview             "/web/build/app/webview/${config}/keymanweb-webview.js" \
```

**Review by jahorton — CHANGES_REQUESTED** [review]

File: common/web/tslib/README.md
Comment (jahorton) [reply common/web/tslib/README.md:1]: ... Hmm.  We went from +32 bytes from the prior PR to +4619 bytes in this one.  It's definitely worth double-checking the build output for duplicated helpers, then.

File: common/web/keyboard-processor/build.sh (hunk 1 above)
Comment (ermshiperete) [reply common/web/keyboard-processor/build.sh]: Makes sense. Done.

File: common/web/keyboard-processor/tests/dom/web-test-runner.config.mjs (hunk 3 above)
Comment (ermshiperete) [reply common/web/keyboard-processor/tests/dom/web-test-runner.config.mjs]: Done.

File: common/web/keyboard-processor/tests/dom/cases/domKeyboardLoader.spec.ts (hunk 2 above)
Comment (ermshiperete) [reply common/web/keyboard-processor/tests/dom/cases/domKeyboardLoader.spec.ts]: This didn't work exactly like this, but I found a similar pattern in `test_utils.ts`, so I applied that.

Comment (jahorton) [reply common/web/keyboard-processor/tests/dom/cases/domKeyboardLoader.spec.ts]: I've submitted a patch PR as #11993 to implement the pattern I suggested.

File: common/web/keyboard-processor/src/text/stringDivergence.ts [hunk 7]
```diff
@@ -1,5 +1,4 @@
-// Future TODO:  import from @keymanapp/common-types... once we no longer need to support ES5.
-import { Uni_IsSurrogate1, Uni_IsSurrogate2 } from '@keymanapp/web-utils';
+import { util } from '@keymanapp/common-types';
```
Comment (jahorton) [inline common/web/keyboard-processor/src/text/stringDivergence.ts:1]: Requiring this style of import for the utility methods prevents treeshaking from working as well as it could.  Having to import the `Uni_IsSurrogate*` functions through a higher-level structure (the collective `export * as util` pattern) results in esbuild importing _all_ functions exported from that structure.

Tracking issue: https://github.com/evanw/esbuild/issues/1420

This can be found within the debug build of web's app/browser when built:

```js
  // ../../../../common/web/types/src/util/util.ts
  var util_exports = {};
  __export(util_exports, {
    BadStringAnalyzer: () => BadStringAnalyzer,
    BadStringType: () => BadStringType,
    CONTAINS_QUAD_ESCAPE: () => CONTAINS_QUAD_ESCAPE,
    MATCH_HEX_ESCAPE: () => MATCH_HEX_ESCAPE,
    MATCH_QUAD_ESCAPE: () => MATCH_QUAD_ESCAPE,
    NFDAnalyzer: () => NFDAnalyzer,
    StringAnalyzer: () => StringAnalyzer,
    UnescapeError: () => UnescapeError,
    Uni_IsSurrogate: () => Uni_IsSurrogate,
    Uni_IsSurrogate1: () => Uni_IsSurrogate1,
    Uni_IsSurrogate2: () => Uni_IsSurrogate2,
    boxXmlArray: () => boxXmlArray,
    describeCodepoint: () => describeCodepoint,
    escapeRegexChar: () => escapeRegexChar,
    escapeStringForRegex: () => escapeStringForRegex,
    hexOcts: () => hexOcts,
    hexQuad: () => hexQuad,
    isOneChar: () => isOneChar,
    isPUA: () => isPUA,
    isValidUnicode: () => isValidUnicode,
    toOneChar: () => toOneChar,
    unescapeOne: () => unescapeOne,
    unescapeOneQuadString: () => unescapeOneQuadString,
    unescapeQuadString: () => unescapeQuadString,
    unescapeString: () => unescapeString,
    unescapeStringToRegex: () => unescapeStringToRegex
  });
  function boxXmlArray(o, x) {
    if (typeof o == "object" && !Array.isArray(o[x])) {
      if (o[x] === null || o[x] === void 0) {
        o[x] = [];
      } else {
        o[x] = [o[x]];
      }
    }
  }
  __name(boxXmlArray, "boxXmlArray");
  var MATCH_HEX_ESCAPE = /\\u{([0-9a-fA-F ]{1,})}/g;
  var CONTAINS_QUAD_ESCAPE = /(?:\\u([0-9a-fA-F]{4})|\\U([0-9a-fA-F]{8}))/;
  var MATCH_QUAD_ESCAPE = new RegExp(CONTAINS_QUAD_ESCAPE, "g");
  var _UnescapeError = class _UnescapeError extends Error {
  };
  __name(_UnescapeError, "UnescapeError");
  var UnescapeError = _UnescapeError;
  function unescapeOne(hex) {
    const codepoint = Number.parseInt(hex, 16);
    return String.fromCodePoint(codepoint);
  }
  __name(unescapeOne, "unescapeOne");
  function unescapeOneQuadString(s) {
    if (!s || !s.match(MATCH_QUAD_ESCAPE)) {
      throw new UnescapeError(`Not a quad escape: ${s}`);
    }
    function processMatch(str, m16, m32) {
      return unescapeOne(m16 || m32);
    }
    __name(processMatch, "processMatch");
    s = s.replace(MATCH_QUAD_ESCAPE, processMatch);
    return s;
  }
  __name(unescapeOneQuadString, "unescapeOneQuadString");
  function unescapeQuadString(s) {
    s = s.replaceAll(MATCH_QUAD_ESCAPE, (quad) => unescapeOneQuadString(quad));
    return s;
  }
  __name(unescapeQuadString, "unescapeQuadString");
  function unescapeString(s) {
    if (!s) {
      return s;
    }
    try {
      let processMatch2 = function(str, matched) {
        const codepoints = matched.split(" ");
        const unescaped = codepoints.map(unescapeOne);
        return unescaped.join("");
      };
      var processMatch = processMatch2;
      __name(processMatch2, "processMatch");
      s = s.replaceAll(MATCH_HEX_ESCAPE, processMatch2);
    } catch (e) {
      if (e instanceof RangeError) {
        throw new UnescapeError(`Out of range while unescaping '${s}': ${e.message}`, { cause: e });
      } else {
        throw e;
      }
    }
    return s;
  }
  __name(unescapeString, "unescapeString");
  function hexQuad(n) {
    if (n < 0 || n > 65535) {
      throw RangeError(`${n} not in [0x0000,0xFFFF]`);
    }
    return n.toString(16).padStart(4, "0");
  }
  __name(hexQuad, "hexQuad");
  function hexOcts(n) {
    if (n < 0 || n > 4294967295) {
      throw RangeError(`${n} not in [0x00000000,0xFFFFFFFF]`);
    }
    return n.toString(16).padStart(8, "0");
  }
  __name(hexOcts, "hexOcts");
  function escapeRegexChar(ch) {
    const code = ch.codePointAt(0);
    if (code <= 65535) {
      return "\\u" + hexQuad(code);
    } else {
      return "\\U" + hexOcts(code);
    }
  }
  __name(escapeRegexChar, "escapeRegexChar");
  var REGEX_SYNTAX_CHAR = /^[\u0000-\u001F\u007F-\u009F{}\[\]\\?|.^$*()/+-]$/;
  function escapeRegexCharIfSyntax(ch) {
    if (REGEX_SYNTAX_CHAR.test(ch) || !isValidUnicode(ch.codePointAt(0))) {
      return escapeRegexChar(ch);
    } else {
      return ch;
    }
  }
  __name(escapeRegexCharIfSyntax, "escapeRegexCharIfSyntax");
  function regexOne(hex) {
    const unescaped = unescapeOne(hex);
    return Array.from(unescaped).map((ch) => escapeRegexCharIfSyntax(ch)).join("");
  }
  __name(regexOne, "regexOne");
  function escapeStringForRegex(s) {
    return s.split("").map((ch) => escapeRegexCharIfSyntax(ch)).join("");
  }
  __name(escapeStringForRegex, "escapeStringForRegex");
  function unescapeStringToRegex(s) {
    if (!s) {
      return s;
    }
    try {
      let processMatch2 = function(str, matched) {
        const codepoints = matched.split(" ");
        const unescaped = codepoints.map(regexOne);
        return unescaped.join("");
      };
      var processMatch = processMatch2;
      __name(processMatch2, "processMatch");
      s = s.replaceAll(MATCH_HEX_ESCAPE, processMatch2);
    } catch (e) {
      if (e instanceof RangeError) {
        throw new UnescapeError(`Out of range while unescaping '${s}': ${e.message}`, { cause: e });
      } else {
        throw e;
      }
    }
    return s;
  }
  __name(unescapeStringToRegex, "unescapeStringToRegex");
  function isOneChar(value) {
    return [...value].length === 1;
  }
  __name(isOneChar, "isOneChar");
  function toOneChar(value) {
    if (!isOneChar(value)) {
      throw Error(`Not a single char: ${value}`);
    }
    return value.codePointAt(0);
  }
  __name(toOneChar, "toOneChar");
  function describeCodepoint(ch) {
    let s;
    const p = BadStringAnalyzer.getProblem(ch);
    if (p != null) {
      s = p;
    } else {
      s = `"${String.fromCodePoint(ch)}"`;
    }
    return `${s} (U+${Number(ch).toString(16).toUpperCase()})`;
  }
  __name(describeCodepoint, "describeCodepoint");
  var BadStringType = /* @__PURE__ */ ((BadStringType2) => {
    BadStringType2["pua"] = "PUA";
    BadStringType2["unassigned"] = "Unassigned";
    BadStringType2["illegal"] = "Illegal";
    BadStringType2["denormalized"] = "Denormalized";
    return BadStringType2;
  })(BadStringType || {});
  var Uni_LEAD_SURROGATE_START = 55296;
  var Uni_LEAD_SURROGATE_END = 56319;
  var Uni_TRAIL_SURROGATE_START = 56320;
  var Uni_TRAIL_SURROGATE_END = 57343;
  var Uni_SURROGATE_START = Uni_LEAD_SURROGATE_START;
  var Uni_SURROGATE_END = Uni_TRAIL_SURROGATE_END;
  var Uni_FD_NONCHARACTER_START = 64976;
  var Uni_FD_NONCHARACTER_END = 65007;
  var Uni_FFFE_NONCHARACTER = 65534;
  var Uni_PLANE_MASK = 2031616;
  var Uni_MAX_CODEPOINT = 1114111;
  var Uni_PUA_00_START = 57344;
  var Uni_PUA_00_END = 63743;
  var Uni_PUA_15_START = 983040;
  var Uni_PUA_15_END = 1048573;
  var Uni_PUA_16_START = 1048576;
  var Uni_PUA_16_END = 1114109;
  function Uni_IsSurrogate1(ch) {
    return ch >= Uni_LEAD_SURROGATE_START && ch <= Uni_LEAD_SURROGATE_END;
  }
  __name(Uni_IsSurrogate1, "Uni_IsSurrogate1");
  function Uni_IsSurrogate2(ch) {
    return ch >= Uni_TRAIL_SURROGATE_START && ch <= Uni_TRAIL_SURROGATE_END;
  }
  __name(Uni_IsSurrogate2, "Uni_IsSurrogate2");
  function Uni_IsSurrogate(ch) {
    return Uni_IsSurrogate1(ch) || Uni_IsSurrogate2(ch);
  }
  __name(Uni_IsSurrogate, "Uni_IsSurrogate");
  function Uni_IsEndOfPlaneNonCharacter(ch) {
    return (ch & Uni_FFFE_NONCHARACTER) == Uni_FFFE_NONCHARACTER;
  }
  __name(Uni_IsEndOfPlaneNonCharacter, "Uni_IsEndOfPlaneNonCharacter");
  function Uni_IsNoncharacter(ch) {
    return ch >= Uni_FD_NONCHARACTER_START && ch <= Uni_FD_NONCHARACTER_END || Uni_IsEndOfPlaneNonCharacter(ch);
  }
  __name(Uni_IsNoncharacter, "Uni_IsNoncharacter");
  function Uni_InCodespace(ch) {
    return ch >= 0 && ch <= Uni_MAX_CODEPOINT;
  }
  __name(Uni_InCodespace, "Uni_InCodespace");
  function Uni_IsValid1(ch) {
    return Uni_InCodespace(ch) && !Uni_IsSurrogate(ch) && !Uni_IsNoncharacter(ch);
  }
  __name(Uni_IsValid1, "Uni_IsValid1");
  function isValidUnicode(start, end) {
    if (!end) {
      return Uni_IsValid1(start);
    } else if (!Uni_IsValid1(end) || !Uni_IsValid1(start) || end < start) {
      return false;
    } else if (start <= Uni_SURROGATE_END && end >= Uni_SURROGATE_START) {
      return false;
    } else if (start <= Uni_FD_NONCHARACTER_END && end >= Uni_FD_NONCHARACTER_START) {
      return false;
    } else if ((start & Uni_PLANE_MASK) != (end & Uni_PLANE_MASK)) {
      return false;
    } else {
      return true;
    }
  }
  __name(isValidUnicode, "isValidUnicode");
  function isPUA(ch) {
    return ch >= Uni_PUA_00_START && ch <= Uni_PUA_00_END || ch >= Uni_PUA_15_START && ch <= Uni_PUA_15_END || ch >= Uni_PUA_16_START && ch <= Uni_PUA_16_END;
  }
  __name(isPUA, "isPUA");
  var _BadStringMap = class _BadStringMap extends Map {
    toString() {
      if (!this.size) {
        return "{}";
      }
      return Array.from(this.entries()).map(([t, s]) => `${t}: ${Array.from(s.values()).map(describeCodepoint).join(" ")}`).join(", ");
    }
  };
  __name(_BadStringMap, "BadStringMap");
  var BadStringMap = _BadStringMap;
  var _StringAnalyzer = class _StringAnalyzer {
    constructor() {
      /** internal map */
      __publicField(this, "m", new BadStringMap());
    }
    /** add a string for analysis */
    add(s) {
      for (const c of [...s]) {
        const ch = c.codePointAt(0);
        const problem = this.analyzeCodePoint(c, ch);
        if (problem) {
          this.addProblem(ch, problem);
        }
      }
    }
    /** internal interface for the result of an analysis */
    addProblem(ch, type) {
      if (!this.m.has(type)) {
        this.m.set(type, /* @__PURE__ */ new Set());
      }
      this.m.get(type).add(ch);
    }
    /** get the results of the analysis */
    analyze() {
      if (this.m.size == 0) {
        return null;
      } else {
        return this.m;
      }
    }
  };
  __name(_StringAnalyzer, "StringAnalyzer");
  var StringAnalyzer = _StringAnalyzer;
  var _BadStringAnalyzer = class _BadStringAnalyzer extends StringAnalyzer {
    /** analyze one codepoint */
    analyzeCodePoint(c, ch) {
      return _BadStringAnalyzer.getProblem(ch);
    }
    /** export analyzer function  */
    static getProblem(ch) {
      if (!isValidUnicode(ch)) {
        return "Illegal" /* illegal */;
      } else if (isPUA(ch)) {
        return "PUA" /* pua */;
      } else {
        return null;
      }
    }
  };
  __name(_BadStringAnalyzer, "BadStringAnalyzer");
  var BadStringAnalyzer = _BadStringAnalyzer;
  var _NFDAnalyzer = class _NFDAnalyzer extends StringAnalyzer {
    analyzeCodePoint(c, ch) {
      const nfd = c.normalize("NFD");
      if (c !== nfd) {
        return "Denormalized" /* denormalized */;
      } else {
        return null;
      }
    }
  };
  __name(_NFDAnalyzer, "NFDAnalyzer");
  var NFDAnalyzer = _NFDAnalyzer;
```

Comment (jahorton) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:1]: If I instead import those functions directly from the original, defining source file...

```js
  // ../../../../common/web/types/src/util/util.ts
  var CONTAINS_QUAD_ESCAPE = /(?:\\u([0-9a-fA-F]{4})|\\U([0-9a-fA-F]{8}))/;
  var MATCH_QUAD_ESCAPE = new RegExp(CONTAINS_QUAD_ESCAPE, "g");
  var Uni_LEAD_SURROGATE_START = 55296;
  var Uni_LEAD_SURROGATE_END = 56319;
  var Uni_TRAIL_SURROGATE_START = 56320;
  var Uni_TRAIL_SURROGATE_END = 57343;
  function Uni_IsSurrogate1(ch) {
    return ch >= Uni_LEAD_SURROGATE_START && ch <= Uni_LEAD_SURROGATE_END;
  }
  __name(Uni_IsSurrogate1, "Uni_IsSurrogate1");
  function Uni_IsSurrogate2(ch) {
    return ch >= Uni_TRAIL_SURROGATE_START && ch <= Uni_TRAIL_SURROGATE_END;
  }
  __name(Uni_IsSurrogate2, "Uni_IsSurrogate2");
```

_**Way**_ less imported code this way.

Alternatively, exporting the two directly from `main`, rather than through the `util.` "namespace" import, would also do the job.

**Review by jahorton — COMMENTED** [review]
The notes added here should not be considered "blockers" - but they may prove material worth a look.  They're a follow-up regarding the noted filesize-increase concerns.

File: common/web/keyboard-processor/src/text/stringDivergence.ts (hunk 7 above)
Comment (jahorton) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:1]: Alternatively... we could just leave the versions defined within `common/web/util` in place, as they're not subject to the import pattern causing the issue.

File: common/web/tslib/README.md
Comment (jahorton) [reply common/web/tslib/README.md:1]: I've been able to confirm that the filesize increase is _not_ due to dropping use of `tslib`.

File: common/web/types/src/main.ts [hunk 8]
```diff
@@ -50,7 +50,7 @@ export { KeymanDeveloperProject, KeymanDeveloperProjectFile, KeymanDeveloperProj
 export * as KpsFile from './package/kps-file.js';
 export * as KmpJsonFile from './package/kmp-json-file.js';
 
-export * as util from './util/util.js';
+export { Uni_IsSurrogate1, Uni_IsSurrogate2 } from './util/util.js';
```
Comment (jahorton) [inline common/web/types/src/main.ts:53]:
```suggestion
export { Uni_IsSurrogate1, Uni_IsSurrogate2 } from './util/util.js';
export * as util from './util/util.js';
```

No reason we can't keep the original export point in place for Developer usages.

**Review by jahorton — APPROVED** [review]
LGTM, though we should probably get @mcdurdin's thoughts on the parts of the changes that affect common/web/types + developer.

File: developer/src/kmc-ldml/src/compiler/messages.ts [hunk 9]
```diff
@@ -1,4 +1,5 @@
-import { util, CompilerErrorNamespace, CompilerErrorSeverity, CompilerMessageSpec as m, CompilerMessageDef as def } from "@keymanapp/common-types";
+import { CompilerErrorNamespace, CompilerErrorSeverity, CompilerMessageSpec as m, CompilerMessageDef as def } from "@keymanapp/common-types";
+import * as util from '@keymanapp/common-types/utils';
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/src/compiler/messages.ts:2]: We shouldn't be using deep imports between the kmc modules -- we are working towards well-defined API surfaces for each module, and this is a definite step backwards. Can we please only use top-level imports.

(Doesn't look like eslint has a [good rule](https://github.com/import-js/eslint-plugin-import/issues/1430) for this at present?)

File: developer/src/kmc-ldml/src/compiler/tran.ts [hunk 10]
```diff
@@ -1,5 +1,6 @@
 import { constants, SectionIdent } from "@keymanapp/ldml-keyboard-constants";
-import { KMXPlus, LDMLKeyboard, CompilerCallbacks, VariableParser, MarkerParser, util } from '@keymanapp/common-types';
+import { KMXPlus, LDMLKeyboard, CompilerCallbacks, VariableParser, MarkerParser } from '@keymanapp/common-types';
+import * as util from '@keymanapp/common-types/utils';
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/src/compiler/tran.ts:3]: Ditto

File: developer/src/kmc-ldml/test/helpers/index.ts [hunk 11]
```diff
@@ -5,7 +5,8 @@ import 'mocha';
 import * as path from 'path';
 import { fileURLToPath } from 'url';
 import { SectionCompiler, SectionCompilerNew } from '../../src/compiler/section-compiler.js';
-import { util, KMXPlus, LDMLKeyboardXMLSourceFileReader, VisualKeyboard, CompilerEvent, LDMLKeyboardTestDataXMLSourceFile, compilerEventFormat, LDMLKeyboard, UnicodeSetParser, CompilerCallbacks } from '@keymanapp/common-types';
+import { KMXPlus, LDMLKeyboardXMLSourceFileReader, VisualKeyboard, CompilerEvent, LDMLKeyboardTestDataXMLSourceFile, compilerEventFormat, LDMLKeyboard, UnicodeSetParser, CompilerCallbacks } from '@keymanapp/common-types';
+import * as util from '@keymanapp/common-types/utils';
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/test/helpers/index.ts:9]: ditto

File: common/web/types/package.json [hunk 12]
```diff
@@ -12,6 +12,10 @@
     ".": {
       "es6-bundling": "./src/main.ts",
       "default": "./build/src/main.js"
+    },
+    "./utils": {
+      "es6-bundling": "./src/util/index.ts",
+      "default": "./build/src/util/index.js"
```
Comment (mcdurdin) [inline common/web/types/package.json:18]: What is the purpose of this? I don't want to see separate exports for the module -- it'd be better to split it into a separate module than to do this.

File: developer/src/kmc-ldml/src/compiler/empty-compiler.ts [hunk 13]
```diff
@@ -1,6 +1,7 @@
 import { SectionIdent, constants } from '@keymanapp/ldml-keyboard-constants';
 import { SectionCompiler } from "./section-compiler.js";
-import { LDMLKeyboard, KMXPlus, CompilerCallbacks, util, MarkerParser } from "@keymanapp/common-types";
+import { LDMLKeyboard, KMXPlus, CompilerCallbacks, MarkerParser } from "@keymanapp/common-types";
+import * as util from '@keymanapp/common-types/utils';
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/src/compiler/empty-compiler.ts:4]: ditto

**Review by mcdurdin — CHANGES_REQUESTED** [review]
A couple of things I am not comfortable with here. The sharing of code common/web/types between Developer and Web seems still to be a little difficult. Perhaps we should have a discussion about that?

File: developer/src/kmc-ldml/src/compiler/messages.ts (hunk 9 above)
Comment (jahorton) [reply developer/src/kmc-ldml/src/compiler/messages.ts:2]: Context: https://github.com/keymanapp/keyman/pull/11881#discussion_r1687319082

Without this, KMW will grow approximately 4 KB in filesize.

Comment (mcdurdin) [reply developer/src/kmc-ldml/src/compiler/messages.ts:2]: I saw the context. I am asking that we don't fix the growth issue by introducing a new problem.

We do _not_ need to use this pattern. We can just export the two functions that KMW needs from common-types/main.ts.

Comment (jahorton) [reply developer/src/kmc-ldml/src/compiler/messages.ts:2]: > We do not need to use this pattern. We can just export the two functions that KMW needs from common-types/main.ts.

I do agree with that; it does feel like the better approach.

File: web/build.sh [hunk 14]
```diff
@@ -39,11 +39,34 @@ builder_describe "Builds engine modules for Keyman Engine for Web (KMW)." \
...
+  build:engine/element-wrappers "/web/build/engine/element-wrappers/lib/index.mjs" \
+  build:engine/main             "/web/build/engine/main/lib/index.mjs" \
+  build:engine/osk              "/web/build/engine/osk/lib/index.mjs" \
+  build:engine/package-cache    "/web/build/engine/package-cache/lib/index.mjs" \
+  build:engine/paths            "/web/build/engine/paths/lib/index.mjs" \
+  build:samples                 "/web/src/samples/simplest/keymanweb.js" \
+  build:tools                   "/web/build/tools/building/sourcemap-root/index.js" \
+  build:test-pages              "/web/build/test-resources/sentry-manager.js"
+
+BUNDLE_CMD="node ${KEYMAN_ROOT}/common/web/es-bundling/build/common-bundle.mjs"
```
[the remark this answers is not in the document]
Comment (ermshiperete) [reply web/build.sh:68]: There are a lot of places, so I'm moving this to a separate issue (#12052).

File: developer/src/kmc-ldml/src/compiler/messages.ts (hunk 9 above)
Comment (ermshiperete) [reply developer/src/kmc-ldml/src/compiler/messages.ts]: Done

File: developer/src/kmc-ldml/src/compiler/tran.ts (hunk 10 above)
Comment (ermshiperete) [reply developer/src/kmc-ldml/src/compiler/tran.ts]: Done

File: developer/src/kmc-ldml/test/helpers/index.ts (hunk 11 above)
Comment (ermshiperete) [reply developer/src/kmc-ldml/test/helpers/index.ts]: Done

File: common/web/types/package.json (hunk 12 above)
Comment (ermshiperete) [reply common/web/types/package.json]: No longer needed. Removed.

File: developer/src/kmc-ldml/src/compiler/empty-compiler.ts (hunk 13 above)
Comment (ermshiperete) [reply developer/src/kmc-ldml/src/compiler/empty-compiler.ts]: Done

File: developer/src/kmc-ldml/src/compiler/messages.ts (hunk 9 above)
Comment (ermshiperete) [reply developer/src/kmc-ldml/src/compiler/messages.ts]: ok, the reason I did it this way was to keep the size of `keymanweb.js` the same. To achieve this I moved the constants to `util/consts.ts`. These didn't get exported in `main.ts`, instead I exported `@keymanapp/common-types/utils`. Just exporting the two functions that KMW needs in addition to `*` didn't help with the size.

So I think we have two options:
- export the consts in `main.ts` again and live with the increased size,
- create a completely separate module for `common-types/utils`.

@mcdurdin @jahorton What do you think? If the second option, what should it be named?

Comment (jahorton) [reply developer/src/kmc-ldml/src/compiler/messages.ts]: It's fine if the constants are the only "extras" along for the ride; their file-size contribution is minimal.  This is what I see on my local build - the _constants_ are pulled in, but they're just four extra lines, and the longest part of each is a variable name that can be substantially minified.

Here's the minified form of what we currently get:

```js
var dQ=55296,gQ=56319,uQ=56320,IQ=57343;function vs(Q){return Q>=dQ&&Q<=gQ}l(vs,"Uni_IsSurrogate1");function Ws(Q){return Q>=uQ&&Q<=IQ}l(Ws,"Uni_IsSurrogate2")
```

That's a total of... 40 characters.  It'd be lovely if we didn't have them, but it's not worth hyper-optimizing.

Comment (jahorton) [reply developer/src/kmc-ldml/src/compiler/messages.ts]: So, to give a concise answer, I pick option 3:  neither.  Leave it as you currently have it.

Comment (mcdurdin) [reply developer/src/kmc-ldml/src/compiler/messages.ts]: I'm not seeing the size issue at present -- I think @jahorton's final [comment above](https://github.com/keymanapp/keyman/pull/11881/files/02170dd0625763c5e3f4a89f19e26ff9dc8a654f#r1696162301) is probably the best answer (and should resolve the current build failures I hope!)

Comment (ermshiperete) [reply developer/src/kmc-ldml/src/compiler/messages.ts]: yes, that works, thanks! Strange that it makes a difference whether or not the consts are defined and exported in the same file...

**Review by mcdurdin — APPROVED** [review]
LGTM

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.80-alpha](https://keyman.com/downloads/releases/alpha/18.0.80)
